policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+            return
+
+        target_key = obj.key
+
+        if action == CMD_ADDED:
+            if target_key not in resource_pool:
+                resource_pool[target_key] = obj
+                _log.info("%s %s added to Calico store" % (kind, target_key))
+            else:
+                _log.error("Tried to Add %s %s, but it was already in bin" % \
+                           (kind, target_key))
+        elif action == CMD_MODIFIED:
+            if target_key in resource_pool:
+                _log.info("Updating %s %s" % (kind, target_key))
+                resource_pool[target_key] = obj
+            else:
+                _log.warning("Tried to Modify %s %s, but it was not in bin. "
+                             "Treating as Addition" % (kind, target_key))
+                self.process_resource(action=CMD_ADDED,
+                                      kind=kind,
+                                      target=target)
+        else:
+            _log.error("Event in process_resource not recognized: %s" % r_type)
+
+    def delete_resource(kind, target):
+        if kind == KIND_NAMESPACE:
+            resource_pool = self.namespaces
+            obj = Namespace(target)
+
+        elif kind == KIND_SERVICE:
+            resource_pool = self.services
+            obj = Service(target)
+
+        elif kind == KIND_POD:
+            resource_pool = self.pods
+            obj = Pod(target)
+
+        elif kind == KIND_ENDPOINTS:
+            resource_pool = self.endpoints
+            obj = Endpoints(target)
+
+        target_key = obj.key
+
+        if target_key in resource_pool:
+            del resource_pool[target_key]
+            _log.info("%s %s deleted from Calico store" % (kind, target_key))
+
+        else:
+            _log.error("Tried to Delete %s %s but it was not in bin" % \
+                       (kind, target_key))
+
+    def match_pod(self, namespace, name):
+        try:
+            return self.pods["%s/%s" % (namespace, name)]
+        except KeyError:
+            _log.error("Pod %s in NS %s not processed" % (name, namespace))
+            return None
+
+    def resync(self):
+        """
+        Updates Calico store with new resource information
+        """
+        for namespace_key, ns in self.changed_namespaces.items():
+            # Create/update Namespace Profiles
+            if ns.create_ns_profile():
+                self.nsResourceVersion = ns.resourceVersion
+                self.namespaces[namespace_key] = ns
+                del self.changed_namespaces[namespace_key]
+
+        for service_key, svc in self.changed_services.items():
+            # Process Services
+            self.svcResourceVersion = svc.resourceVersion
+            self.services[service_key] = svc
+            del self.changed_services[service_key]
+
+        for pod_key, po in self.changed_pods.items():
+            # Apply Namespace policy to new/updated pods
+            if po.apply_ns_policy():
```

Comment:
I'm not a super big fan of these methods which perform an action (such as apply ns policy) and then return True / False based on their success - it isn't very Pythonic.

Methods like "apply_ns_profile" which perform a very specific task are also a bit of a turn off, since they're not very flexible or re-usable.

I'd expect this API to look more like this:
```
for pod_key, pod in self.changed_pods.iteritems():
  namespace = self.namespaces.get(pod.namespace)
  if not namespace:
    # Namespace for this pod does not yet exist - skip.  It will retain the default REJECT_ALL profile.
    _log.warning(...)
    continue
  
  # The namespace exists.  Set the profiles - this also removes any existing profiles on the pod.
  ns_profile = namespace.profile
  pod.set_profiles([ns_profile])

  # Update our internal state
  self.podResourceVersion = pod.resourceVersion
  self.pods[pod.key] = pod
  del self.changed_pods[pod.key]
```